FB2: fix various issues

poire-z opened the pull request:
Fix most issues with FB2 noticed from https://github.com/koreader/koreader/issues/6344#issuecomment-655936850 and later comments.

`CSS: adds text-align-last: -cr-right-if-not-first private keywords` 
To be used with FB2 `<v>` https://github.com/koreader/koreader/issues/6344#issuecomment-657347443

`CSS: adds "-cr-only-if: fb2-document"`
Might allow to target some style tweaks to only EPUB or FB2, and avoid some conflicts. https://github.com/koreader/koreader/issues/6344#issuecomment-658273890

`FB2 footnotes: fix spacing between number and text`
Have consistent (1 space, fixed width) spacing between FB2 footnote number and text.

`ldomTextCollector: guess block boundaries better`
Avoid spurious newlines in text selection in various context (FB2 footnotes, ruby...)

(Changes not super tested with formats other than FB2, so bumping this might wait till after next release.)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/crengine/357)
<!-- Reviewable:end -->

Frenzie commented:
I'm pretty much just waiting for https://github.com/koreader/koreader/pull/6364 and for myself to write some release notes. :-)

virxkane commented:
@poire-z About CSS: adds text-align-last: -cr-right-if-not-first private keywords. Why you think that it's right?
In the browser, this HTML is rendered differently.
[page2_2.html.zip](https://github.com/koreader/crengine/files/4946153/page2_2.html.zip)
I can't find in fb2 specification any alignment instructions: https://github.com/gribuser/fb2/blob/master/FictionBook.xsd

poire-z commented:
Follow the link to the issue where we discussed it (that I usually mention in the first port of a PR), https://github.com/koreader/koreader/issues/6344#issuecomment-657347443 .

Summary:
With `text-align-last: right`, we behave as a browser.
But @hius07 and I find that ugly - and there is no `text-align-last` specified keyword to have it like we expect verses/poetry to be in a paperbook - so I added this one.
(Note that I fixed `text-align-last: right` to actually behave as browsers do with https://github.com/koreader/crengine/commit/5b28e2bd38a94e5b67aaf6c569be1e843072aedd #345 - before, we behave just like with this added `text-align-last: -cr-right-if-not-first` - so you're actually getting back how `<v>` used to be rendered.)

The FB2 specs specify structure/hierarchy, not rendering - that is left to the implementation where we have a bit of freedom to express our aesthetic preferences :)

virxkane commented:
> Follow the link to the issue where we discussed it (that I usually mention in the first port of a PR), [koreader/koreader#6344 (comment)](https://github.com/koreader/koreader/issues/6344#issuecomment-657347443) .

Of course, I read this thread. 

> we expect verses/poetry to be in a paperbook

This is more clearer. But it depends on book, is there any standard or generally accepted rules for the design of poetry in paper books?
Too many non-standard items have already been added: -cr-ignore-if-dom-version-greater-or-equal, -cr-hint, -cr-only-if, -cr-list-item-final and newest "-cr-*-if-not-first". I would be very careful when introducing new non-standard keywords, otherwise at some point you might find that we have some new ugly standard. For example, some newbies are trying to customize fb2.css, epub.css, luckily they know css, but what do they know about "-cr-right-if-not-first", "-cr-hint"? It is clear that sometimes you cannot do without them, but the fewer the better.
Or soon you will have to rename *.css to *.crss :)

poire-z commented:
Do you think I added them for pleasure ?! :)

Most were added to ensure compatibility with previous renderings/DOM, or allow previous/bogus behaviour. Sometimes because we need to - to not lose past highlights, before the normalized XPointers update, sometimes because I'm unsure of the new stuff and I think some people migh prefer the old behaviour.

On KOReader, we have the ability to inject small snippets of CSS, known as "Style tweaks". Some use these -cr properties and keywords. Users are not really expected to know about them. Most are only internal stuff, used to drive the engine - and they sometimes allowed to remove ugly hardcoded stuff  like `-cr-hint: footnote-inpage` allowed to remove FB2 hardcoded footnotes handling, and have them available for EPUB too.
Same for `-cr-right-if-not-first` - no standard keyword allows for the expected poetry alignment behaviour. I could just have not cared, or I could have hardcoded `<v>` handling - but these keywords make that less hardcoded and re-usable - and fix a hole in the specs.

So, never a pleasure - most often a cross to bear :)
(and sorry to see you carry a bit of it and scratch your head when merging :)

(I'll probably add soon other -cr-hint to allow/forbid glyph overflows, cf https://github.com/koreader/crengine/pull/355#issuecomment-656758176 - because there is actually some code that handles that - italic overflows, `getAdditionalCharWidth()` - contrary to most browsers - I could just drop that code - but I think some users might one day be happy to be able to get the previous non-standard behaviour in some contexts, including me :) And may be it was introduced for some Russian/Cyrillic glyphs characteristics, I suspect cyrillic having more tall/square glyphs might trigger these glyphs overriding each other more easily than latin text, so that's why may be all these italic checks were added - pinged you in https://github.com/koreader/crengine/pull/355#issuecomment-657013772 .)

pkb commented:
As I understood poire-z, they are needed to switch (on/off) either previous/bogus or non standard behaviour. You can make them UI options.

poire-z commented:
> this is implemented to be more similar to the behavior of the old version of the program? 

What "this" ? All of the stuff? `-cr-right-if-not-first` ? my glyph overflows ideas ?

> Is it possible to use gDOMVersionRequested for this purpose

Well, logically, gDOMVersionRequested is for DOM incompatibilities. All these hint stuff are more about rendering, nothing to do with the DOM.
And we usually want to migrate users to use the latest DOM version (I expect, with the xpointers migration, that we'll never have to increase gDOMVersion - but never say never :) So, it would be ugly to force using an older gDOMVersionRequested to get good-looking-but-wrong `text-align-last: right` support with FB2 just because they used to be specified as `v { text-align-last: right; }` - but have to use a newer gDOMVersionRequested for EPUB, because there, publisher might use `DIV.lookAtMyStyle { text-align-last: right; }` to actually explicitely expect what you get with browsers with your page2_2.html.zip .

> so as not to introduce new non-standard keywords?

I understand your concern about non-standard keywords - but in this case, it's actually the right non-ugly way to fix the issue. (Well, what's ugly is its name, `right-if-not-first` :) if english had a word for that, `fleft` or `fright` , I would have used it instead :)

> But it depends on book, is there any standard or generally accepted rules for the design of poetry in paper books?

FB2 does not specify rendering, it just specifies that verses in poems are to be wrapped in `<v>` - and we (actually, _you_, Russian FB2 readers :) could have left it unstyled - but _you_ decided to have verses a bit nicer looking.

In EPUBs, there's no such thing. Publisher can wrap poems in anything, and style that how they like it.

No paper book would render a 4-verses poem like this I guess (**A**)
<kbd>![image](https://user-images.githubusercontent.com/24273478/87937312-c78b7f00-ca94-11ea-9561-db20f4bcbf9f.png)</kbd>

We all expect to see them either just wrapped like normal lines of dialog, or better like this (**B**):
<kbd>![image](https://user-images.githubusercontent.com/24273478/87937372-e427b700-ca94-11ea-8f8a-5969c33ee28c.png)</kbd>

CoolReader since 2010 used to render them like **B** (and I think it's an aesthetic choice by buggins - I don't think it's an implementation error of the text-align-last specs - so, I think it's an assumed divergence/hijack of the CSS specs to get the pleasant rendering - just like `display: run-in`, in some specs at some point, was used for FB2 footnotes).
CoolReader can no more get it like **B** since my #345.
**And no EPUB publisher can achieve B with normal CSS**, as no CSS classic properties or keywords allow to get **B** (or I would have used/implemented them if I had found any! :)

So, as CSS don't allow us to get **B**, we could just forget about the nice looking **B**.

Or we could add a keyword that the CSS spec designers did not care about introducing - which I did - and have **B** again.

And if I ever read a poem in some EPUB that is laid out like normal lines of dialog - I, with KOReader, can add a user style tweak to have it rendered as `B`:
`div.frenchPoetryVerse { text-align-last: -cr-right-if-not-first; }`

I hope that makes my journey clearer to you :)

virxkane commented:
@poire-z Please, don't overreact. Of course I understood what the problem was, you didn't have to repeat yourself. I am not nitpicking, but trying to understand your decision or find another in a constructive dialogue.  If I don’t find it, I’ll take yours. Yes, I understand gDOMVersionRequested  is better not to touch.
Sorry, if it seemed rude, I do not know English well.

poire-z commented:
Thanks, but don't worry - I'm not overreacting, just over-explaining.
After all, it's easier for me to explain stuff I spent hours thinking about (even if it takes a bit of time writting it) - than trying to understand somebody else thoughts and code :)
(And so that you don't have to spend another several days rethinking/reimplementing something that non-essential :)

poire-z commented:
> That is, after all, this is implemented to be more similar to the behavior of the old version of the program? Is it possible to use gDOMVersionRequested for this purpose, so as not to introduce new non-standard keywords?

(OK, I guess you were asking that more generally than I understood it :)

Some of them are implemented to be more similar to the behavior of the old version of the program.
But not all.
Here's how we went at it, if I can make some sense of these 3 years of hacking/fixing/improving crengine:

- When the old stuff is bad or bogus, and there's no need to keep it: I just fix/drop it :)

- When the old stuff is bad, bogus (like the wrong `display:` values for some elements) but needed to not lose past users highlights, I made it dependant on gDomVersionRequested. Same for the block rendering flags, I kept the "legacy" behaviour, while implementing the "enhanced" one (with various flags we can set from frontend - because there too, rendering just "as browsers" may not be best for books and eReaders - ex: browsers allow width overflows, but CoolReader and KOReader don't allow scrolling the page horizontally - so we can drop some flags to be not as perfect as browsers, but friendlier to eReaders)

- When the old stuff is nice looking (better than browsers) or can be useful in some cases, but different from the standard, it makes sense to keep it available: `right-if-not-first` is such a case to me. Would feed sad to drop it (and make @hius07 disapointed) just to be "like browsers". We had something good, I don't want to drop it to be "standard". The same may apply to italic glyph overflows.

- When some new good idea might not be browser-conforming, or might have some bad effect on the default rendering of publishers books, or is a question of taste - and I don't have a strong feeling that my taste is worth killing all other tastes :): I try to make it an option or toggable in some way:

- Either by a new crengine toggle (like 140044bd2), but that is global and quite a pain to add (lvdocviewprops...) and needs some frontend changes
- Either via CSS, where it's easier to handle, can be targeted to some specific elements like TABLEs, or FB2 footnotes and verses, and does not need UI changes - and that's easy to tackle with KOReader and its user style tweaks.

`-cr-hint` is indeed my prefered way for such ideas, and I added a few that make them optionnal, user toggable and targetable, while allowing us to be like-browsers by default (browsers don't have "in-page" footnotes).
https://github.com/koreader/crengine/blob/686c7daee2aec8d3d4dee11f9c07258e4cbbf9c2/crengine/src/lvstsheet.cpp#L1792-L1824

Sometimes, an idea might fit better in an existing CSS property: aligning the last line of a verse fits better as `text-align-last: -cr-right-if-not-first` than as `-cr-hint: align-last-line-right`.

Btw, you picked html5.css for HTML.
I hate html5.css :) I added it because it's nice to have it available as the starting point for users who like it, or because some EPUBs expect it to be the base, and have stylesheet to just add stuff to this base, that would not work with epub.css.
html5.css drops parapgraph indent, adds some vertical margins between paragraphs. But that's what browsers do. But it's ugly on plain simple HTML or EPUBs.
Our epub.css is still our default, because it makes 90% of the normal EPUBs or simple HTML files look as we expect it, or as we are used to see it with KOReader or CoolReader - and we still use it with epub and html files.

Once again, we don't need to be just as browsers, when we can be better - sometimes we're better by default, or standard by default , but we have tweaks, hints and toggles to allow freedom on all that.
I reckon it's a lot to take in when you're picking to merge into CoolReader - moreover when you don't have KOReader's style tweaks facilities to witness how all these CSS are nice, and some of them might make no sense.

virxkane commented:
> * When the old stuff is nice looking (better than browsers) or can be useful in some cases, but different from the standard, it makes sense to keep it available: `right-if-not-first` is such a case to me. Would feed sad to drop it (and make @hius07 disapointed) just to be "like browsers". We had something good, I don't want to drop it to be "standard".

There is even nothing to discuss here, of course you need to return it as it was in the old version.

> No paper book would render a 4-verses poem like this I guess (A)

Of course, I have not seen such books. Let me write some obvious things too.
Alexander Pushkin (https://fr.wikipedia.org/wiki/Alexandre_Pouchkine): 
![097A2028-web](https://user-images.githubusercontent.com/36960933/87968559-aac16c80-cad1-11ea-9d09-6f2ab705f725.JPG)

![IMG_20200720_174237-web](https://user-images.githubusercontent.com/36960933/87968518-98dfc980-cad1-11ea-847f-59108a43cfd7.jpg)
You can clearly see that the left alignment is used, except for the quote by Vasily Zhukovsky, but the "v" tag is definitely not applicable to it.


> CoolReader since 2010 used to render them like **B**

To be more precise 3.0.46-11 (2011-06-20).
Ok, trying this version
File test1.fb2 ([test1.fb2.zip](https://github.com/koreader/crengine/files/4949243/test1.fb2.zip) )
![cr3 0 46-test1](https://user-images.githubusercontent.com/36960933/87975272-72735b80-cadc-11ea-972a-5d251d168478.png)
In this version fb2.css contains such record about tag "v": v { text-align: left; text-align-last: right; text-indent: 1em hanging }
"text-align-last: right" not work. At what point Biggins fixed it, I did not find out.
You can himsef test in https://sourceforge.net/projects/crengine/files/CoolReader3/cr3-3.0.46/
Now test text rendering with real poetry Alexander Pushkin "Mon portait" (1814) 
[Pushkin_A-S-Mon portrait.fb2.zip](https://github.com/koreader/crengine/files/4949261/Pushkin_A-S-Mon.portrait.fb2.zip)
![cr3 0 46-mon-portrait](https://user-images.githubusercontent.com/36960933/87975465-b8c8ba80-cadc-11ea-907b-964d5e6a63cb.png)

Ok, go next CoolReader 3.2.11 (before merging with your crengine fork):
File test1.fb2:
![cr3 2 11-test1](https://user-images.githubusercontent.com/36960933/87970459-d72ab800-cad4-11ea-917c-f0508a6bf933.png)
File Pushkin_A-S-Mon.portrait.fb2:
![cr3 2 11-mon-portrait](https://user-images.githubusercontent.com/36960933/87970510-e7db2e00-cad4-11ea-9989-97df7c51facc.png)

Go next, koreader-2014-04-122
File test1.fb2:
![koreader-2014 04 122-test1](https://user-images.githubusercontent.com/36960933/87971183-faa23280-cad5-11ea-89d2-a35340e199d1.png)
File Pushkin_A-S-Mon.portrait.fb2:
![koreader-2014 04 122-mon-portrait](https://user-images.githubusercontent.com/36960933/87971203-01c94080-cad6-11ea-910b-833f95d2d336.png)

Go next, my current "not-yet-merged" branch
File test1.fb2
![cr3-virxkane-current-test1](https://user-images.githubusercontent.com/36960933/87970716-312b7d80-cad5-11ea-9225-0f8af2067476.png)
File Pushkin_A-S-Mon.portrait.fb2:
![cr3-virxkane-current-mon-portrait](https://user-images.githubusercontent.com/36960933/87970741-38528b80-cad5-11ea-9b71-e8be33ea088d.png)
Ok looks wrong.

Go next, koreader-2020.06-60-g7849d58_2020-07-19
File test1.fb2:
![koreader-2020 06-60-g7849d58_2020-07-19-test1](https://user-images.githubusercontent.com/36960933/87971438-65536e00-cad6-11ea-86ef-905d540d360a.png)
File Pushkin_A-S-Mon.portrait.fb2:
![koreader-2020 06-60-g7849d58_2020-07-19-mon-portrait](https://user-images.githubusercontent.com/36960933/87971460-6c7a7c00-cad6-11ea-8162-54371ba59bee.png)

Yes, looks wrong, but why don't we remove the "text-align-last" property from tag "v".

Go next, my current "not-yet-merged" branch, property text-align-last removed
File test1.fb2
![cr3-virxkane-current-test1-skip-text-align-last](https://user-images.githubusercontent.com/36960933/87971821-ffb3b180-cad6-11ea-8673-2138effe7c46.png)
Ok, logicaly wrong, but  who said it "should be right aligned"?
File Pushkin_A-S-Mon.portrait.fb2:
![cr3-virxkane-current-mon-portrait-skip-text-align-last](https://user-images.githubusercontent.com/36960933/87971998-3f7a9900-cad7-11ea-8741-5d0322b7b254.png)
How do you say, voilà.

And again, but why don't we remove the "text-align-last" property from tag "v"?
Could you provide me with an example where the properties of the tag "v" are "text-align-last: right" really need?
Lines that are too long to fit on one line and wrap to the second line?
Does anyone really want the wrapped line to be right-justified?
May set "text-align-last: center"?

poire-z commented:
> Could you provide me with an example where the properties of the tag "v" are "text-align-last: right" really need?
**Lines that are too long to fit on one line and wrap to the second line?**

Yes, it's **exactly and only just for this case**: when a verse is too long and needs to be wrapped.

Without any `text-align-last`:
<kbd>![image](https://user-images.githubusercontent.com/24273478/87979638-a6e21a00-cad2-11ea-9ed3-1ddba1bbb6c0.png)</kbd>

With `text-align-last: -cr-right-if-not-first`:
<kbd>![image](https://user-images.githubusercontent.com/24273478/87979710-c5e0ac00-cad2-11ea-99f1-bb0fbe86a009.png)</kbd>

With `text-align-last: center`:
<kbd>![image](https://user-images.githubusercontent.com/24273478/87979890-0b04de00-cad3-11ea-93d8-a8b7239ad1af.png)</kbd>

I don't read much poetry, and in the few paperbooks I have, I can't find any where a wrap was needed (only one, which uses hanging indent for the 2nd line).
But from what I remember from school, I think I've seen some verses laid out like on the 2nd screenshot above (that's why it feels not unnatural to me, and why I considered me killing it unfortunate, and felt the need to repair this).

Anway, I'm not the one to argue with about the _default_ rendering of FB2 verses :) Have that discussion with @hius07 and other FB2 readers.

But I'll argue that, given how cheap it was adding `-cr-right-if-not-first`, and that at least 2 persons like it (hius07 and buggins), it would be sad to drop the possibility for them to have it (with style tweaks or an edited fb2.css) for the sake of not abusing private CSS properties (which is not a disease ! :)

hius07 commented:
> Lines that are too long to fit on one line and wrap to the second line?

Exactly, see Homerus (lines 4 and 8):

![1](https://user-images.githubusercontent.com/62179190/88007748-5214c280-cb17-11ea-8306-1590c0add0ee.jpg)

virxkane commented:
@hius07 Thank you very much, I searched for a paper book with hyphenated lines of verses, but to no avail.

virxkane commented:
There is one solution, but unfortunately it does not applicable, since it does not meet the standard, but for some reason it works in Chromium.
[test4_3.html.zip](https://github.com/koreader/crengine/files/4952349/test4_3.html.zip)
https://www.w3.org/TR/css-pseudo-4/#selectordef-first-line
https://developer.mozilla.org/en-US/docs/Web/CSS/::first-line
In Firefox this don't work.

poire-z commented:
https://github.com/koreader/koreader/issues/6344#issuecomment-657472168

Also mentionned in the link pointed to by @Frenzie at https://github.com/koreader/koreader/issues/6344#issuecomment-661212090:
https://lists.w3.org/Archives/Public/www-style/2014Jun/0167.html
```
  fantasai: We also have cases where people would want the
            ::first-line aligned for poetry or something.
  TabAtkins: Can you apply text-align to the ::first-line?
  fantasai: I don't know. I think you could.
  TabAtkins: They don't cascade together.
  fantasai: The cascading is a bit of a problem.
  fantasai: I think it is.
```
I don't want to understand the details of cascading and precedence and all the complexities these people mention (I just understand it might be a bad idea :)

Frenzie commented:
@virxkane See https://github.com/koreader/koreader/issues/6344#issuecomment-657459505

It has always been per-spec, but apparently Firefox has managed to push through their weird restrictions that have annoyed me since 2003?!

https://www.w3.org/TR/CSS1/#the-first-line-pseudo-element
https://www.w3.org/TR/CSS2/selector.html#first-line-pseudo
https://drafts.csswg.org/selectors-3/#first-line

Frenzie commented:
@poire-z Pseudoselectors don't cascade, that's all they're saying. So you can't just set it on html or body and call it a day.

Frenzie commented:
@virxkane My bad, apparently they already managed to sneak that into CSS3.

> The ::first-line pseudo-element is similar to an inline-level element, but with certain restrictions. The following CSS properties apply to a ::first-line pseudo-element: font properties, color property, background properties, ‘word-spacing’, ‘letter-spacing’, ‘text-decoration’, ‘text-transform’, ‘line-height’. UAs may apply other properties as well.

Frenzie commented:
But anyway, that's unrelated to the issue at hand. It'd be a reasonably elegant hack, but no more than that. `last-if-not-first` is a pretty obvious oversight imho. ^_^

virxkane commented:
Apparently there is no better variant, will use this one. Although, of course, not good variant.
Thanks to all.
